File: C++/008. String to Integer (atoi).cpp
#include <iostream>
#include <climits>

class Solution {
public:
    int myAtoi(std::string str) {
        int k;
        //删除前面所有空白字符
        for (k = 0; k < int(str.size()) && str[k] == ' '; k++);
        str = str.substr(k);
        //判断符号
        int symbol = 1;
        if (str.size() > 0 && str[0] == '+')
            str = str.substr(1);
        else if (str.size() > 0 && str[0] == '-') {
            symbol = -1;
            str = str.substr(1);
        }
        //截取有效字符
        for (k = 0; k < int(str.size()) && str[k] >= '0' && str[k] <= '9'; k++);
        str = str.substr(0, k);
        //转换成数字
        long res = 0;
        for (k = 0; k < int(str.size()); k++) {
            res *= 10;
            res += str[k] - '0';
            if (symbol * res > INT32_MAX)
                return INT32_MAX;
            if (symbol * res < INT32_MIN)
                return INT32_MIN;
        }
        return int(symbol * res);
    }
};

File: Go/008. String to Integer (atoi).go
const INT_MAX = 2147483647
const INT_MIN = -2147483648

func myAtoi(str string) int {
	var i int
	for i = 0; i < len(str) && str[i] == ' '; i++ {}
	str = str[i : ]
	var symbol int64 = 1
	if len(str) > 0 && str[0] == '+' {
		str = str[1 :]
	} else if len(str) > 0 && str[0] == '-' {
		symbol = -1
		str = str[1 :]
	}
	for i = 0; i < len(str) && str[i] >= '0' && str[i] <= '9'; i++ {}
	str = str[0 : i]
	var res int64 = 0
	for i = 0; i < len(str); i++ {
		res *= 10
		res += int64(str[i] - '0')
		if symbol * res > INT_MAX {
			return INT_MAX
		}
		if symbol * res < INT_MIN {
			return INT_MIN
		}
	}
	return int(symbol * res)
}